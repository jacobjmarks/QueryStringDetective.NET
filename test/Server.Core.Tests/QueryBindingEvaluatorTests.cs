namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "",
        "q=",
        "q=c",
        "q=foo",
        "q=a&q=b&q=c",
        "q=null",
        "q=0",
        "q=1",
        "q=-1",
        "q=true",
        "q=false",
        "q=yes",
        "q=no",
        "q=1&q=2&q=3",
        "q=1&q=2&q=.3",
        "q=-1&q=0&q=1",
        $"q={sbyte.MinValue}",
        $"q={sbyte.MaxValue}",
        $"q={byte.MaxValue}",
        $"q={short.MinValue}",
        $"q={short.MaxValue}",
        $"q={ushort.MaxValue}",
        $"q={int.MinValue}",
        $"q={int.MaxValue}",
        $"q={long.MinValue}",
        $"q={long.MaxValue}",
        $"q={ulong.MaxValue}",
        $"q={Uri.EscapeDataString($"{double.MinValue}")}",
        $"q={Uri.EscapeDataString($"{double.MaxValue}")}",
        $"q={decimal.MinValue}",
        $"q={decimal.MaxValue}",
        $"q={Uri.EscapeDataString($"{float.MinValue}")}",
        $"q={Uri.EscapeDataString($"{float.MaxValue}")}",
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?" + queryString);

        var content = $"Input: {queryString}"
            + "\n\nBindings:\n";

        foreach (var result in results)
        {
            content += "\n" + result.Type;
            content += "\n  " + string.Join("\n  ", result.Results
                .Select(r => $"{r.Key}\n    {(
                    r.Value.IsErroneous
                    ? $"e = {r.Value.Error.Message} ({r.Value.Error.Detail})"
                    : $"r = {r.Value.Result}")}"));

            content += "\n";
        }

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}