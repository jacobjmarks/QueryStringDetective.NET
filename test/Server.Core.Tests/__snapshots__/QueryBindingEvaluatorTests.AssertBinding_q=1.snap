Input: q=1

Bindings:

string
  MinimalApis
    r = "1"
  Controllers
    r = "1"

string?
  MinimalApis
    r = "1"
  Controllers
    r = "1"

string[]
  MinimalApis
    r = ["1"]
  Controllers
    r = ["1"]

char
  MinimalApis
    r = "1"
  Controllers
    r = "1"

char?
  MinimalApis
    r = "1"
  Controllers
    r = "1"

char[]
  MinimalApis
    r = ["1"]
  Controllers
    r = ["1"]

int
  MinimalApis
    r = 1
  Controllers
    r = 1

int?
  MinimalApis
    r = 1
  Controllers
    r = 1

int[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

double
  MinimalApis
    r = 1
  Controllers
    r = 1

double?
  MinimalApis
    r = 1
  Controllers
    r = 1

double[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

float
  MinimalApis
    r = 1
  Controllers
    r = 1

float?
  MinimalApis
    r = 1
  Controllers
    r = 1

float[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

decimal
  MinimalApis
    r = 1
  Controllers
    r = 1

decimal?
  MinimalApis
    r = 1
  Controllers
    r = 1

decimal[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

bool
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool q" from "1".)
  Controllers
    e = 400 Bad Request (The value '1' is not valid.)

bool?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<bool> q" from "1".)
  Controllers
    e = 400 Bad Request (The value '1' is not valid.)

bool[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool[] q" from "1".)
  Controllers
    e = 400 Bad Request (The value '1' is not valid.)

uint
  MinimalApis
    r = 1
  Controllers
    r = 1

uint?
  MinimalApis
    r = 1
  Controllers
    r = 1

uint[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

long
  MinimalApis
    r = 1
  Controllers
    r = 1

long?
  MinimalApis
    r = 1
  Controllers
    r = 1

long[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

ulong
  MinimalApis
    r = 1
  Controllers
    r = 1

ulong?
  MinimalApis
    r = 1
  Controllers
    r = 1

ulong[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

sbyte
  MinimalApis
    r = 1
  Controllers
    r = 1

sbyte?
  MinimalApis
    r = 1
  Controllers
    r = 1

sbyte[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

short
  MinimalApis
    r = 1
  Controllers
    r = 1

short?
  MinimalApis
    r = 1
  Controllers
    r = 1

short[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

ushort
  MinimalApis
    r = 1
  Controllers
    r = 1

ushort?
  MinimalApis
    r = 1
  Controllers
    r = 1

ushort[]
  MinimalApis
    r = [1]
  Controllers
    r = [1]

byte
  MinimalApis
    r = 1
  Controllers
    r = 1

byte?
  MinimalApis
    r = 1
  Controllers
    r = 1

byte[]
  MinimalApis
    r = "AQ=="
  Controllers
    e = 400 Bad Request (The value '1' is not valid.)
